src/server/scene/surface_stack.cpp

```diff
@@ -425,61 +425,75 @@ void ms::SurfaceStack::raise(SurfaceSet const& ss)
-            int first_index = -1;
-            int second_index = -1;
-            std::vector<std::shared_ptr<Surface>> first_move;
-            std::vector<std::shared_ptr<Surface>> second_move;
+            // The goal is to swap the first set with the second set such that their Z-order is swapped.
+
+            // First, find the start and end of the range that we want to swap
+            auto swap_begin = layer.begin();
+            auto swap_end = layer.end();
+            size_t num_first_found = 0;
+            size_t num_second_found = 0;
+            bool first_to_front = true;
             for (auto it = layer.begin(); it != layer.end(); it++)
             {
+                // Once we've found them all, then we have the whole range to sort across
+                if (num_first_found == first.size() && num_second_found == second.size())
+                {
+                    swap_end = it;
+                    break;
+                }
+
+                // Find the start position and count how many we've found
                 if (first.count(*it))
                 {
-                    if (first_index < 0)
-                        first_index = it - layer.begin();
-                    first_move.push_back(*it);
+                    if (!num_first_found && !num_second_found)
+                        swap_begin = it;
+
+                    num_first_found++;
                 }
                 else if (second.count(*it))
                 {
-                    if (second_index < 0)
-                        second_index = it - layer.begin();
-                    second_move.push_back(*it);
+                    if (!num_first_found && !num_second_found)
+                    {
+                        first_to_front = false;
+                        swap_begin = it;
+                    }
+
+                    num_second_found++;
                 }
             }
 
-            // Break early if there is nothing to swap here
-            if (second_index < 0 || first_index < 0)
-                continue;
-
-            // Delete the swapped elements  from the list
-            for (auto it = layer.begin(); it != layer.end();)
+            // Finally, move the to_front items to the front of the group and the to_back to the back of the group
+            auto to_front = first_to_front ? first : second;
+            auto to_back = first_to_front ? second : first;
+            std::stable_sort(swap_begin, swap_end, [to_front, to_back](std::weak_ptr<Surface> s1, std::weak_ptr<Surface> s2)
             {
-                auto found_first = std::find(first_move.begin(), first_move.end(), *it);
-                auto found_second = std::find(second_move.begin(), second_move.end(), *it);
-                if (found_first != first_move.end() || found_second != second_move.end())
-                     layer.erase(it);
+                if (to_front.count(s1))
+                {
+                    return -1;
+                }
```

Comment:
Note that the result of the compare functor is contextually converted to bool: returning `-1`, `1` and `0` is confusing.

Also, for a strict weak ordering, if both `s1` and `s2` are in the same equivalence class (e.g. `to_front.count(s1)` and `to_front.count(s2)`) then the result should be `false`.